### ISSUE TITLE
[LOWPRI][FUTURE]Support empty variant type, and empty record type.

### ISSUE THREAD

**Issue by jordwalke** [facebook/reason#1597]
@objmagic added these: https://twitter.com/objmagic/status/927026404331479041

We have one issue with syntactic real estate for empty records. Currently we use `{}` to represent empty Object, but I don't really think it's used that frequently - we merely needed `{}` to parse to something so that when people are writing dummy function bodies, it doesn't error. I think `{}` would be a better candidate for empty record, now that it's supported.
It will take some time for these empty variant types/record types to work their way into the OCaml ecosystem so support for this is low pri right now, but it's good to be aware of it.

**Comment by objmagic** [comment]
A working branch is [here](https://github.com/ocaml/ocaml/compare/trunk...objmagic:empty-record-variants?expand=1). It still needs some polish and I need to send it to experienced maintainer for a preview. Below is an example

For the following program:

```OCaml
type o = A
type t = | ;;
let f (x:t) = match x with _ -> . ;;
type m = { };;
let x = { }
type n = m
let y:n = x
module X = struct type g = { }  end
let g = { }
let n = let open X in { }
let h = n
```

The signature will be:
```
type o = A
type t = |
val f : t -> 'a
type m = { }
val x : m
type n = m
val y : n
module X : sig type g = { } end
val g : m
val n : X.g
val h : X.g
```

**Comment by jordwalke** [comment]
I'm still confused about how this is useful in practice. Care to TL;DR?

**Comment by objmagic** [comment]
Here is an example of [why empty variants could be useful](https://caml.inria.fr/mantis/view.php?id=7583#c18077).

For empty record, as Yallop mentioned in the above Mantis ticket, compiling empty Protobuf enum could be one example.

I admit that the real-world use cases are pretty rare. So the main purpose of these additions are to make language more consistent.

**Comment by reynir** [comment]
I'm using an empty type to mark constructors in a variant as not-yet-implemented. When I'm ready to implement the constructor I change it from `Foo of empty` to `Foo of int` (say), and the compiler will give me a bunch of errors to guide me in implementing the Foo-constructor. At least that's the idea. Not sure if this is a good approach or not.

**Comment by objmagic** [comment]
It seems empty record is unlikely to be allowed. Jacques showed his arguments here: https://github.com/ocaml/ocaml/pull/1536#issuecomment-353221867

**Comment by anmonteiro** [comment]
Empty variants are implemented in #2487
